src/unittest/test_collision.cpp

```diff
@@ -60,149 +62,194 @@ namespace {
 		UASSERTEQ_F(va.X, ve.X); UASSERTEQ_F(va.Y, ve.Y); UASSERTEQ_F(va.Z, ve.Z); \
 	} while (0)
 
+#define fpos(x,y,z) (BS * v3f(x, y, z))
+
 
 ////////////////////////////////////////////////////////////////////////////////
+TEST_CASE("Test axis aligned collision with unit cube.", "[collision]")
+{
+	f32 bx = GENERATE(-3.0f, -2.0f, -1.0f, 0.0f, 1.0f, 2.0f, 3.0f);
+	f32 by = GENERATE(-3.0f, -2.0f, -1.0f, 0.0f, 1.0f, 2.0f, 3.0f);
+	f32 bz = GENERATE(-3.0f, -2.0f, -1.0f, 0.0f, 1.0f, 2.0f, 3.0f);
+
+	INFO("Testing with static cube at (" << bx << ", " << by << ", " << bz << ").");
+
+	aabb3f s{bx, by, bz, bx+1.0f, by+1.0f, bz+1.0f};
+
+	// The following set of tests is for boxes translated in the -X direction
+	// from the static cube, possibly with additional offsets.
+	SECTION("Given a unit cube translated by -2 units on the x-axis, "
+			"when it moves 1 unit per step in the +x direction for 1 step, "
+			"then it should collide on the X axis within epsilon of 1 step.")
+	{
+		aabb3f m{bx-2.0f, by, bz, bx-1.0f, by+1.0f, bz+1.0f};
+		f32 dtime = 1.0f;
+		CHECK(0 == axisAlignedCollision(s, m, v3f{1.0f, 0.0f, 0.0f}, &dtime));
+		CHECK(std::fabs(static_cast<double>(dtime) - 1.0) < EPSILON);
+	}
+
+	SECTION("Given a unit cube translated by -2 units on the x-axis, "
+			"when it moves 1 unit per step in the -x direction for 1 step, "
+			"then it should never collide.")
+	{
+		aabb3f m{bx-2.0f, by, bz, bx-1.0f, by+1.0f, bz+1.0f};
+		f32 dtime = 1.0f;
+		CHECK(-1 == axisAlignedCollision(s, m, v3f{-1.0f, 0.0f, 0.0f}, &dtime));
+	}
+
+
+	SECTION("Given a unit cube translated by -2 units on the x-axis "
+			"and 1.5 units on the y-axis, "
+			"when it moves 1 unit per step in the +x direction for 1 step, "
+			"then it should never collide.")
+	{
+		aabb3f m{bx-2.0f, by+1.5f, bz, bx-1.0f, by+2.5f, bz+1.0f};
+		f32 dtime = 1.0f;
+		CHECK(-1 == axisAlignedCollision(s, m, v3f{1.0f, 0.0f, 0.0f}, &dtime));
+	}
+
+	SECTION("Given a 0.5x2x1 cube translated by -2 units on the x-axis "
+			"and -1.5 units on the y-axis, "
+			"when it moves 0.5 units per step in the +x direction "
+			"and 0.1 units per step in the +y direction for 3 steps, "
+			"then it should collide on the X axis within epsilon of 3 steps.")
```

Comment:
having throughout but hard to read textual description that duplicates the test contents looks pretty useless to me

Reply:
Yeah, this is pretty bad. I will think about how to describe the intent of the test better.

Reply:
I took a shot at it. Is there anything else I should do to improve them?